### DISCUSSION TITLE
Keyboard reacts to music

### DISCUSSION THREAD

**Discussion by ItsVince1234** [4JX/L5P-Keyboard-RGB#35] [Ideas]
Would it be possible for the keyboard to react to music?

**Comment by 4JX** [comment]
Originally did not have much hope for such an effect, but it seems [its actually possible](https://github.com/dheijl/swyh-rs/blob/master/src/utils/audiodevices.rs#L65L119) to capture outputted audio, so the answer is "probably yes".

That being said, do you have any preference on how it'd visualize it? It could change intensity, color, be a bar (left to right / right to left). I'll probably just end up doing a slight UI rework before doing more than one of the variations.

**Reply by HyperN00B** [reply]
I think flashing the lights in the selected colors with different intensity depending on the music could work.

**Reply by fa1riyi** [reply]
I have been always wondering if you can make 4 zones change their colors seperately (no effect of left to right or right to left). For music, can you alter the speed of color changing base on tempo of music? (not sure if we can capture tempo of music), and the higher tone of sound, the lighter of color, and the lower tone of sound, the darker of color, and this should be based on Hz of sound I think.

**Comment by aarmn** [comment]
I think I can work on implementing this feature, but Im a novice in Rust so I may make it on python, also working with music streams is hugely platform dependent and I found very few libs to support multiple OS and multiple libs in Lin (pulseaudio and ALSA), also its not a promise!!!!!

**Reply by 4JX** [reply]
Hey sorry for the late response on this one, you can try based on my current approach, see:
https://github.com/4JX/L5P-Keyboard-RGB/blob/8b88c8992bf1c847e90f86e560ab6f01e6df5c9f/src/keyboard_manager.rs#L359-L417

Its almost there prototype wise, just haven't figured out why it sets everything to 0 when you add that line.
* Tested it on my Linux machine, you'll need to use `pavucontrol` and in the `Recording` tab, while the `Sound` effect is active, set the _capture from_ to "Monitor of X".

**Reply by Daniel0012** [reply]
how do i add the effect?
